File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.logging.SimpleFormatter;

class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        List<Object> list = new ArrayList<>();
        Date date1 = new Date();
        Calendar calToday = Calendar.getInstance();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy.MM.dd");
        try {
            date1 = sdf.parse(today);
        } catch (ParseException e) {
            throw new RuntimeException(e);
        }
        calToday.setTime(date1);

        for(int i=0;i<privacies.length;i++){
            String term = privacies[i].split(" ")[1];
            int count = 0;
            for(String item : terms){
                String[] items = item.split(" ");
                if(items[0].equals(term)) {
                    count = Integer.parseInt(items[1]);
                    break;
                }
            }
            Calendar cal = Calendar.getInstance();

            Date date = null;
            try {
                date = sdf.parse(privacies[i].split(" ")[0]);
            } catch (ParseException e) {
                throw new RuntimeException(e);
            }
            cal.setTime(date);

            cal.add(Calendar.MONTH, count);
            cal.add(Calendar.DATE, -1);
            if(cal.get(Calendar.DATE) > 28){
                cal.set(Calendar.DATE, 28);
            }
            if(cal.before(calToday)){

                list.add(i+1);
            }
        }
        int[] answer = new int[list.size()];
        for(int i=0;i<list.size();i++){
            answer[i] = (Integer) list.get(i);
        }

        return answer;
        
    }    

}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
from collections import defaultdict
def to_value(today):
    year, month, day = today.split(".")
    return int(year) * 12 * 28 + int(month) * 28 + int(day)

def solution(today, terms, privacies):
    answer = []
    today = to_value(today)
    term_data = defaultdict(int)
    for term in terms:
        t, v = term.split(" ")
        term_data[t] = int(v) * 28

    for i, privacy in enumerate(privacies):
        privacy_today, privacy_type = privacy.split(" ")
        p_today_value = to_value(privacy_today)
        p_term_value = term_data[privacy_type]
        if p_today_value + p_term_value <= today:
            answer.append(i + 1)

    return answer